src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -192,6 +192,11 @@ public VisitArgumentResult(VisitResult visitResult, Optional<LocalState> stateFo
         /// </summary>
         private PooledDictionary<BoundAwaitableValuePlaceholder, (BoundExpression AwaitableExpression, VisitResult Result)>? _awaitablePlaceholdersOpt;
 
+        // Note: at the moment, we have two kinds of placeholders. Those that we visit first (and store the result from and substitute later)
+        // and those that we visit after subsitution.
+        // Ideally, we should be able to align on the first kind of design, then we could remove the second kind.
+        private PooledDictionary<BoundValuePlaceholderBase, BoundExpression>? _placeholdersToUnvisitedExpressionOpt;
```

Comment:
I'm very concern about this. If we're re-visiting BoundExpressions, we could be changing the recorded nullability for those expressions, which will affect the public API, and violating assumptions of the snapshot recorder. We could also end up double-reporting nullability issues inside the replaced content, which to the user just results in duplicate warnings on the same syntax. I don't think this is the correct solution for placeholders: we should instead be recording the result from the original visit of the BoundExpression and that as the result of visiting the placeholder.

Reply:
> If we're re-visiting BoundExpressions

I do not think we are.

Reply:
I talked a bit with Aleksey offline. I'll leave a comment where I think we can add an assert, and I think the comment on this should be updated with a link to [https://github.com/dotnet/roslyn/issues/57855](https://github.com/dotnet/roslyn/issues/57855). I didn't get the sense from the comment on this member that there was any real desire to not leave this general machinery in the walker, as we have many of these types of comments in the compiler that have similar lack of urgency associated with them.